### ISSUE TITLE
Shopify Order refunds are not parsed correctly

### ISSUE THREAD

**Issue by manavkohli** [airbytehq/airbyte#7577]
## Enviroment
- **Airbyte version**:  0.30.8-alpha
- **OS Version / Instance**: AWS EC2
- **Deployment**:  Docker
- **Source Connector and version**: Shopify 0.1.18
- **Destination Connector and version**: N/A - present in both Postgres and BigQuery
- **Severity**:  Critical
- **Step where error happened**: Sync job

## Current Behavior
Order refunds are nested within orders from the Shopify Order API. Airbyte loads the refund as a nested column in the Order table, but also creates a separate table for refunds. Unfortunately, we've noticed that a significant number of refunds are not being unrolled correctly. We verified this by inspecting the refunds table, comparing it to tables loaded by Fivetran for the same store, and identifying Orders that had non-empty refund JSON column values and no corresponding refund records. We also verified that the refunds loaded by Airbyte did not accurately reflect the data in the Shopify dashboard.

## Expected Behavior
Refunds should be extracted and loaded for all orders.

## Logs
*If applicable, please upload the logs from the failing operation. 
For sync jobs, you can download the full logs from the UI by going to the sync attempt page and 
clicking the download logs button at the top right of the logs display window.*

<details>
<summary>LOG</summary> 

```

replace this with
your long log
output here

```

</details>

## Steps to Reproduce
1. Create and authenticate a Shopify source
2. Create and authenticate a Postgres destination
3. Create a connection for the two

## Are you willing to submit a PR?

Remove this with your answer.

**Comment by sherifnada** [comment]
@manavkohli trying to zero in on where the issue is. Do you know if the correct data is present in the raw tables, or the nested/non-normalized columns in the "orders" table?

**Comment by ChristopheDuong** [comment]
Yes, I agree, we should make sure whether the issue is:
1. the connector did not write all the refunds data (nested inside the orders table) during replication into the _airbyte_raw tables
2. the refund data is complete in JSON format but is not "unrolled" properly into a separate "Refund" table by normalization.

Does it sound by this following sentence that we are leaning more towards point 2?

> We verified this by inspecting the refunds table, and identifying Orders that had non-empty refund JSON column values and no corresponding refund records

is that correct?

Would it be possible to share some json examples (anonymized refunds records in orders rows) that were not properly extracted?

**Comment by manavkohli** [comment]
Yep can clean up some of the order records and send over! And yes I think case 2. To verify this we loaded the data into BigQuery, then executed a query that:
1. Read the refund column in the orders table
2. Applied a regex to extract the quantity of items refunded for each order refund line in the JSON column

We confirmed that the data returned via this method correctly matched the number of refunds loaded by Fivetran/were available in the dashboard.

**Comment by ChristopheDuong** [comment]
Could you also let us know which sync mode were you using? Was it Deduped History?

**Comment by vitaliizazmic** [comment]
@manavkohli hello, i'm trying to reproduce issue. Could you please provide which sync mode you use and provide some example of incorrect unrolled refunds.

**Comment by manavkohli** [comment]
We were using incremental/append_dedup

**Comment by augan-rymkhan** [comment]
We could not reproduce this bug, so I asked the client to try the sync with the latest version of the connector 0.1.25.
So, I am waiting for the response from him.

**Comment by manavkohli** [comment]
@augan-rymkhan @VasylLazebnyk @vitaliizazmic  @ChristopheDuong  I think I figured out the issue here. Basically, the Shopify Orders API may only return orders *that do not have refunds*. I see the same issue with Fivetran. The way the Shopify API exposes refunds is through the Orders API (i.e. you need an order ID to fetch the refund) - therefore the refunds will not be accessible if the order is not. I'm trying to investigate right now whether this is the case, and how else refunded orders would be exposed. Or whether this is a setting on the Shopify store.

**Comment by augan-rymkhan** [comment]
@manavkohli Can you give more details or instructions how to manually set up this condition ?  I mean _I looked at orders that had a refund but no matching transaction. This suggests that the refund was entered but not processed by a POS_

**Comment by sherifnada** [comment]
@manavkohli are you open to sharing an example of a record which has a refund but no transactions? 

My hypothesis is that there are missing fields in the record which are causing normalization not to flatten the record into the appropriate table.

**Comment by manavkohli** [comment]
@sherifnada can do - want to setup some time to debug this live?

**Comment by manavkohli** [comment]
You can also verify - just find records where `updated_at != created_at` and a refund exists

**Comment by manavkohli** [comment]
@sherifnada any help here?

**Comment by marcosmarxm** [comment]
@manavkohli can you send me a Slack message so we can schedule some time to debug it?

**Comment by bazarnov** [comment]
@marcosmarxm @manavkohli @sherifnada 
The issue was hidden within the structure of the nested entities. Record `Orders` could contain nested `Refunds` entity, however, the records inside are not sorted, therefore they could be skipped by `filter_records_newer_than_state` method, because of `ShopifySubstream` class works via `stream_slices` and `stream_state` is updated after each slice. 

If the record contains `cursor_field < current state` it will be skipped. The nested sorting resolves the issue. Here is the fix:
https://github.com/airbytehq/airbyte/pull/17777

**Comment by sherifnada** [comment]
nice find @bazarnov !
